src/requestSchemas.ts

```diff
@@ -0,0 +1,157 @@
+    if (issue.params?.type === 'NotImplemented') {
+      throw new NotImplementedError(issue.message ?? ctx.defaultError);
+    } else {
+      throw new BadRequestError(issue.message ?? ctx.defaultError, issue.params?.serverErrorCode);
+    }
+  } else if (issue.code === z.ZodIssueCode.unrecognized_keys) {
+    throw new BadRequestError(issue.message ?? ctx.defaultError, constants.ISSUE.CODE.VALUE);
+  }
+  throw new BadRequestError(issue.message ?? ctx.defaultError);
+};
+
+z.setErrorMap(customErrorMap);
+
+const hasIdentifyingInfo = (args: Record<string, any>) => args.id || args.url || args.identifier;
+
+/**
+ * Returns a function that checks if any unsupported params are present, then runs the
+ * other passed in functions in sequence. Each catchFunction is expected to check for
+ * invalid input and call ctx.addIssue if invalid input is detected. Any added issue will
+ * trigger an error to be thrown.
+ */
+export function catchInvalidParams(
+  catchFunctions: ((val: Record<string, any>, ctx: z.RefinementCtx) => void)[],
+  unsupportedArgs?: string[]
+) {
+  return (val: Record<string, any>, ctx: z.RefinementCtx) => {
+    const includedUnsupported = Object.keys(val).filter((v: string) => unsupportedArgs?.includes(v));
+    if (includedUnsupported.length > 0) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        params: { type: 'NotImplemented' },
+        message: `Parameters: ${includedUnsupported.join(', ')} are not yet supported on this server`
+      });
+    } else {
+      catchFunctions.forEach(func => func(val, ctx));
+    }
+  };
+}
+
+/**
+ * Checks that either id, url, or identifier parameter is included. Adds an issue to the ctx
+ * that triggers a BadRequest to be thrown if not.
+ */
+export function catchMissingIdentifyingInfo(val: Record<string, any>, ctx: z.RefinementCtx) {
+  if (!hasIdentifyingInfo(val)) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.custom,
+      params: { serverErrorCode: constants.ISSUE.CODE.REQUIRED },
+      message: 'Must provide identifying information via either id, url, or identifier parameters'
+    });
+  }
+}
+
+export const IdentifyingParameters = z
+  .object({
+    id: z.string(),
+    url: z.string().url(),
+    version: z.string(),
+    identifier: z.string()
+  })
+  .partial();
+
+export const PackageArgs = IdentifyingParameters.extend({
+  capability: z.string(),
+  offset: z.string(),
+  count: z.string(),
+  'system-version': z.string(),
+  'check-system-version': z.string(),
+  'force-system-version': z.string(),
+  manifest: z.string(),
+  'include-dependencies': z.string(),
+  'include-components': z.string()
+})
+  .partial()
+  .strict()
+  .superRefine(catchInvalidParams([catchMissingIdentifyingInfo], UNSUPPORTED_PACKAGE_ARGS));
+
+export const DataRequirementsArgs = IdentifyingParameters.extend({
+  expression: z.string(),
+  parameters: z.string(),
+  manifest: z.string(),
+  periodStart: z.string(),
+  periodEnd: z.string(),
```

Comment:
can probably refine these further to ensure they adhere to the FHIR date regex? e.g.

`z.string().regex(/([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)(-(0[1-9]|1[0-2])(-(0[1-9]|[1-2][0-9]|3[0-1]))?)/)` (taken from FHIR spec)

could probably pull the regex out into a variable and reuse it

Reply:
Yeah good thinking. Zod allows a .string().dateTime() type, but I chose not to use it because it rejects just `2019-01-01` style dates. regex seems like a great solution